File: joints-commons/src/main/scala/org/joints/commons/ScriptUtils.scala
package org.joints.commons

import java.util
import javax.script._

import jdk.nashorn.api.scripting.ScriptUtils
import org.apache.commons.lang3.StringUtils
import org.apache.logging.log4j.{LogManager, Logger}

import scala.collection.JavaConverters._

object ScriptUtils {
    private val log: Logger = LogManager.getLogger(classOf[ScriptUtils])

    private val sem: ScriptEngineManager = new ScriptEngineManager

    private val scriptEnginePool: ThreadLocal[util.Map[String, ScriptEngine]] =
        ThreadLocal.withInitial(() => new util.HashMap[String, ScriptEngine]())

    log.info("initialize ScriptEngineManager ")

    for (sef <- sem.getEngineFactories.asScala) {
        log.info(
            s"""
               |engine: {\n\t${sef.getEngineName},
               |version:\t ${sef.getExtensions},
               |mime_types:\t ${sef.getMimeTypes},
               |names:\t ${sef.getNames},
               |language_name:\t ${sef.getLanguageName},
               |language_version:\t ${sef.getLanguageVersion}
               |}
             """.stripMargin)
        sef.getMimeTypes.forEach((mimeType: String) => sem.registerEngineMimeType(mimeType, sef))
        sef.getExtensions.forEach((extension: String) => sem.registerEngineMimeType(extension, sef))
    }

    def getScriptEngineByMimeType(mimeType: String): ScriptEngine = {
        if (StringUtils.isBlank(mimeType)) return null
        val mineTypeAndScriptEngines: util.Map[String, ScriptEngine] = scriptEnginePool.get

        mineTypeAndScriptEngines.computeIfAbsent(mimeType, (_mimeType: String) => {
            val se: ScriptEngine = Option(sem.getEngineByMimeType(_mimeType)).getOrElse(sem.getEngineByExtension(_mimeType))
            if (se == null) log.error("not script engine found for mime _type: " + mimeType)
            se
        })
    }

    def getCompiledScript(mimeType: String, scriptStr: String): CompiledScript = {
        val errorMessage: String = s"ScriptEngine: ${mimeType} can't compile script: \n${scriptStr}"
        try {
            val se: ScriptEngine = getScriptEngineByMimeType(mimeType)
            if (se.isInstanceOf[Compilable]) {
                val compiler: Compilable = se.asInstanceOf[Compilable]
                return compiler.compile(scriptStr)
            }
            log.error(errorMessage)
        } catch {
            case e: ScriptException => log.error(errorMessage, e)
        }
        null
    }
}

File: joints-commons/src/main/java/org/joints/commons/ScriptUtils.java
package org.joints.commons;

import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.script.*;
import java.util.HashMap;
import java.util.Map;

public class ScriptUtils {
	private static final Logger log = LogManager.getLogger(ScriptUtils.class);

	public static CompiledScript getCompiledScript(String mimeType, String scriptStr) {
		String errorMessage = String.format("ScriptEngine: %s can't compile script: \n%s", mimeType, scriptStr);
		try {
			ScriptEngine se = getScriptEngineByMimeType(mimeType);
			if (se instanceof Compilable) {
				Compilable compiler = (Compilable) se;
				return compiler.compile(scriptStr);
			}
			log.error(errorMessage);
		} catch (ScriptException e) {
			log.error(errorMessage, e);
		}
		return null;
	}

	public static ScriptEngine getScriptEngineByMimeType(String mimeType) {
		if (StringUtils.isBlank(mimeType)) return null;
		
		Map<String, ScriptEngine> mineTypeAndScriptEngines = scriptEnginePool.get();
		return mineTypeAndScriptEngines.computeIfAbsent(mimeType, (_mimeType) -> {
			ScriptEngine se = sem.getEngineByMimeType(_mimeType);
			if (se == null) {
				se = sem.getEngineByExtension(_mimeType);
			}
			if (se == null) {
				log.error("not script engine found for mime type: " + mimeType);
			}
			return se;
		});
	}

	private static ThreadLocal<Map<String, ScriptEngine>> scriptEnginePool = ThreadLocal.withInitial(HashMap<String, ScriptEngine>::new);
	
	protected static ScriptEngineManager sem = new ScriptEngineManager();
	
	static {
		log.info("initialize ScriptEngineManager ");
		for (ScriptEngineFactory sef : sem.getEngineFactories()) {
			final String _str = "engine: { \n\tname: " + sef.getEngineName()
					+ ", \n\tversion: " + sef.getEngineVersion()
					+ ", \n\textensions: " + sef.getExtensions()
					+ ", \n\tmime_types: " + sef.getMimeTypes()
					+ ", \n\tnames: " + sef.getNames()
					+ ", \n\tlanguage_name: " + sef.getLanguageName()
					+ ", \n\tlanguage_version: " + sef.getLanguageVersion() + "\n}\n";
			
			log.info(_str);

			sef.getMimeTypes().forEach(mimeType -> sem.registerEngineMimeType(mimeType, sef));
			sef.getExtensions().forEach(extension -> sem.registerEngineMimeType(extension, sef));
		}
	}

}
